# Preprocess

In [35]:
import numpy as np
import pandas as pd

### 데이터 호출

In [36]:
_ROOT_DIR = "./../../"

In [37]:
# data call
# 기본 전처리는 train, test 모두에 적용
def call_data():
    ret_train_df = pd.read_csv(_ROOT_DIR+ './data/raw/cell2celltrain.csv')
    ret_test_df = pd.read_csv(_ROOT_DIR+ './data/raw/cell2cellholdout.csv')


    return ret_train_df, ret_test_df

In [38]:
train_df, test_df = call_data()

In [39]:
# 기본정보 확인
def check_datas(df):
    print("------Head------")
    display(df.head())
    print("------Info------")
    display(df.info())
    print("------Describe------")
    display(df.describe())
    print("------중복값 확인------")
    print(df.duplicated().sum())

In [40]:
# train_df
check_datas(train_df)

------Head------


,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,ReferralsMadeBySubscriber,IncomeGroup,OwnsMotorcycle,AdjustmentsToCreditRating,HandsetPrice,MadeCallToRetentionTeam,CreditRating,PrizmCode,Occupation,MaritalStatus
0,3000002,Yes,24.00,219.0,22.0,0.25,0.0,0.0,-157.0,-19.0,...,0,4,No,0,30,Yes,1-Highest,Suburban,Professional,No
1,3000010,Yes,16.99,10.0,17.0,0.00,0.0,0.0,-4.0,0.0,...,0,5,No,0,30,No,4-Medium,Suburban,Professional,Yes
2,3000014,No,38.00,8.0,38.0,0.00,0.0,0.0,-2.0,0.0,...,0,6,No,0,Unknown,No,3-Good,Town,Crafts,Yes
3,3000022,No,82.28,1312.0,75.0,1.24,0.0,0.0,157.0,8.1,...,0,6,No,0,10,No,4-Medium,Other,Other,No
4,3000026,Yes,17.14,0.0,17.0,0.00,0.0,0.0,0.0,-0.2,...,0,9,No,1,10,No,1-Highest,Other,Professional,Yes


------Info------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51047 entries, 0 to 51046
Data columns (total 58 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   CustomerID                 51047 non-null  int64  
 1   Churn                      51047 non-null  object 
 2   MonthlyRevenue             50891 non-null  float64
 3   MonthlyMinutes             50891 non-null  float64
 4   TotalRecurringCharge       50891 non-null  float64
 5   DirectorAssistedCalls      50891 non-null  float64
 6   OverageMinutes             50891 non-null  float64
 7   RoamingCalls               50891 non-null  float64
 8   PercChangeMinutes          50680 non-null  float64
 9   PercChangeRevenues         50680 non-null  float64
 10  DroppedCalls               51047 non-null  float64
 11  BlockedCalls               51047 non-null  float64
 12  UnansweredCalls            51047 non-null  float64
 13  CustomerCareCalls          51

None

------Describe------


,CustomerID,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,DroppedCalls,...,Handsets,HandsetModels,CurrentEquipmentDays,AgeHH1,AgeHH2,RetentionCalls,RetentionOffersAccepted,ReferralsMadeBySubscriber,IncomeGroup,AdjustmentsToCreditRating
count,5.104700e+04,50891.000000,50891.000000,50891.000000,50891.000000,50891.000000,50891.000000,50680.000000,50680.000000,51047.000000,...,51046.000000,51046.000000,51046.000000,50138.000000,50138.000000,51047.000000,51047.000000,51047.000000,51047.000000,51047.000000
mean,3.201957e+06,58.834492,525.653416,46.830088,0.895229,40.027785,1.236244,-11.547908,-1.191985,6.011489,...,1.805646,1.558751,380.545841,31.338127,21.144142,0.037201,0.018277,0.052070,4.324524,0.053911
std,1.169056e+05,44.507336,529.871063,23.848871,2.228546,96.588076,9.818294,257.514772,39.574915,9.043955,...,1.331173,0.905932,253.801982,22.094635,23.931368,0.206483,0.142458,0.307592,3.138236,0.383147
min,3.000002e+06,-6.170000,0.000000,-11.000000,0.000000,0.000000,0.000000,-3875.000000,-1107.700000,0.000000,...,1.000000,1.000000,-5.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,3.100632e+06,33.610000,158.000000,30.000000,0.000000,0.000000,0.000000,-83.000000,-7.100000,0.700000,...,1.000000,1.000000,205.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,3.201534e+06,48.460000,366.000000,45.000000,0.250000,3.000000,0.000000,-5.000000,-0.300000,3.000000,...,1.000000,1.000000,329.000000,36.000000,0.000000,0.000000,0.000000,0.000000,5.000000,0.000000
75%,3.305376e+06,71.065000,723.000000,60.000000,0.990000,41.000000,0.300000,66.000000,1.600000,7.700000,...,2.000000,2.000000,515.000000,48.000000,42.000000,0.000000,0.000000,0.000000,7.000000,0.000000
max,3.399994e+06,1223.380000,7359.000000,400.000000,159.390000,4321.000000,1112.400000,5192.000000,2483.500000,221.700000,...,24.000000,15.000000,1812.000000,99.000000,99.000000,4.000000,3.000000,35.000000,9.000000,25.000000


------중복값 확인------
0


In [41]:
# test_df
check_datas(test_df)

------Head------


,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,ReferralsMadeBySubscriber,IncomeGroup,OwnsMotorcycle,AdjustmentsToCreditRating,HandsetPrice,MadeCallToRetentionTeam,CreditRating,PrizmCode,Occupation,MaritalStatus
0,3000006,NaN,57.49,483.0,37.0,0.25,23.0,0.0,532.0,51.0,...,0,5,No,1,150,No,5-Low,Other,Other,No
1,3000018,NaN,55.23,570.0,72.0,0.00,0.0,0.0,38.0,0.0,...,0,6,No,2,80,No,1-Highest,Other,Professional,No
2,3000034,NaN,97.34,1039.0,50.0,4.95,420.0,0.0,198.0,23.3,...,0,4,No,3,10,No,3-Good,Suburban,Crafts,Yes
3,3000070,NaN,35.59,153.0,30.0,0.00,16.0,0.0,30.0,7.3,...,0,4,No,1,200,No,1-Highest,Other,Other,No
4,3000074,NaN,55.27,1213.0,50.0,0.74,0.0,1.3,169.0,1.0,...,0,3,No,0,10,No,1-Highest,Suburban,Other,No


------Info------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 58 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   CustomerID                 20000 non-null  int64  
 1   Churn                      0 non-null      float64
 2   MonthlyRevenue             19940 non-null  float64
 3   MonthlyMinutes             19940 non-null  float64
 4   TotalRecurringCharge       19940 non-null  float64
 5   DirectorAssistedCalls      19940 non-null  float64
 6   OverageMinutes             19940 non-null  float64
 7   RoamingCalls               19940 non-null  float64
 8   PercChangeMinutes          19865 non-null  float64
 9   PercChangeRevenues         19865 non-null  float64
 10  DroppedCalls               20000 non-null  float64
 11  BlockedCalls               20000 non-null  float64
 12  UnansweredCalls            20000 non-null  float64
 13  CustomerCareCalls          20

None

------Describe------


,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,Handsets,HandsetModels,CurrentEquipmentDays,AgeHH1,AgeHH2,RetentionCalls,RetentionOffersAccepted,ReferralsMadeBySubscriber,IncomeGroup,AdjustmentsToCreditRating
count,2.000000e+04,0.0,19940.000000,19940.000000,19940.000000,19940.000000,19940.000000,19940.000000,19865.000000,19865.000000,...,20000.000000,20000.000000,20000.000000,19665.000000,19665.000000,20000.000000,20000.000000,20000.00000,20000.000000,20000.000000
mean,3.201925e+06,NaN,58.899712,525.912136,46.946640,0.893764,40.240722,1.186765,-9.052806,-1.237246,...,1.816200,1.569550,379.550450,31.469413,21.192321,0.036500,0.017000,0.04775,4.359000,0.051250
std,1.165204e+05,NaN,43.564203,530.825481,24.083351,2.116538,95.740506,6.848963,249.609021,36.638292,...,1.348669,0.914223,255.552901,22.050726,23.882983,0.204132,0.138967,0.24119,3.134009,0.353312
min,3.000006e+06,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,-2518.000000,-459.000000,...,1.000000,1.000000,-5.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000
25%,3.100522e+06,NaN,33.720000,159.000000,30.000000,0.000000,0.000000,0.000000,-82.000000,-7.200000,...,1.000000,1.000000,202.000000,0.000000,0.000000,0.000000,0.000000,0.00000,1.000000,0.000000
50%,3.201338e+06,NaN,48.665000,365.000000,45.000000,0.250000,2.000000,0.000000,-4.000000,-0.300000,...,1.000000,1.000000,331.000000,36.000000,0.000000,0.000000,0.000000,0.00000,5.000000,0.000000
75%,3.304237e+06,NaN,70.882500,718.000000,60.000000,0.990000,40.000000,0.300000,66.000000,1.600000,...,2.000000,2.000000,513.000000,48.000000,44.000000,0.000000,0.000000,0.00000,7.000000,0.000000
max,3.399986e+06,NaN,635.060000,7668.000000,350.000000,67.810000,3472.000000,488.800000,3222.000000,895.600000,...,28.000000,16.000000,1823.000000,98.000000,99.000000,4.000000,4.000000,5.00000,9.000000,14.000000


------중복값 확인------
0


## 범주형 feature 확인

In [42]:
# 범주형 데이터 unique 값 확인
def check_unique_values(df):
    for column in df.columns:
        if df[column].dtype == 'object':
            unique_values = df[column].unique()
            print(f"Column: {column}, Unique Values: {unique_values}")

In [43]:
check_unique_values(train_df)

Column: Churn, Unique Values: ['Yes' 'No']
Column: ServiceArea, Unique Values: ['SEAPOR503' 'PITHOM412' 'MILMIL414' 'OKCTUL918' 'OKCOKC405' 'SANMCA210'
 'SLCSLC801' 'LOULOU502' 'KCYKCK913' 'KCYNEW316' 'KCYKCM816' 'DENDEN303'
 'PHICTR610' 'OKCLRK501' 'OMADES515' 'SANAUS512' 'KCYWIC316' 'INDIND317'
 'SLCPRO801' 'OMALNC402' 'NSHNSH615' 'OMAOMA402' 'PHXTUC520' 'DALDAL214'
 'DALDTN940' 'DALFTW817' 'SANSAN210' 'NYCMAN917' 'NOLKEN504' 'MINMIN612'
 'NNYROC716' 'PHIARD610' 'NYCQUE917' 'DENBOU303' 'BOSBOS617' 'PHXPHX602'
 'NNYBUF716' 'DETTOL419' 'NNYSYR315' 'STLSTL314' 'PHIPHI215' 'SEABLV425'
 'MIAMIA305' 'PHICHC215' 'SEASEA206' 'BIRBIR205' 'MIADFD954' 'LOUNAL812'
 'MIADEL561' 'MIAFTL954' 'SEASPO509' 'SEATAC253' 'SFRSAC916' 'MIANDA305'
 'MINSTP612' 'MIAHWD954' 'NYCBRO917' 'PHIJEN215' 'OMACDR319' 'MIAWPB561'
 'SEAOLY360' 'PITBUT412' 'SEAEVE425' 'PHIMER609' 'SEACDA208' 'HARHAR860'
 'SFRSFR415' 'SFROAK510' 'SFRSCL408' 'NNYALB518' 'STLCOL618' 'NYCNEW201'
 'SFRPAL650' 'DETDET313' 'DETROS810' 'NYCNEW9

- Service Area
    - unique 값이 지나치게 많아 학습에 영향이 갈 수 있음
    - 비슷한 내용을 담은 PrizmCode가 존재하므로 제거
- Homeownership
    - 자가 소유 여부에 대해서 알고 모르고는 도움이 안됨
    - 방해 요소로 판단하고 제거
- HandsetPrice
    - 수치형 데이터에 가끔 'Unknown'이 끼어 있어 범주형으로 계산됨
    - Unknown을 결측치로 간주
- MaritalStatus
    - 'No', 'Yes'에 'Unknown'이 끼어 있음
    - Unknown을 결측치로 간주
- 그 외 범주형에는 인코딩 필요

## 기본 전처리

In [44]:
# Churn (Yes/No) -> 1/0
# test_df에는 churn이 NaN로 되어있음, 수정 불필요
def preprocess_churn(train_df):
    train_df['Churn'] = train_df['Churn'].map({'Yes': 1, 'No': 0})

    return train_df

In [45]:
# 그 외 인코딩
def preprocess_encoding(train_df, test_df):
    # Yes/No -> 1/0
    binary_cols = [
    'ChildrenInHH', 'HandsetRefurbished', 'HandsetWebCapable', 'TruckOwner', 
    'RVOwner', 'BuysViaMailOrder', 'RespondsToMailOffers', 'OptOutMailings', 
    'NonUSTravel', 'OwnsComputer', 'HasCreditCard', 'NewCellphoneUser', 
    'NotNewCellphoneUser', 'OwnsMotorcycle', 'MadeCallToRetentionTeam'
    ]

    for col in binary_cols:
        train_df[col] = train_df[col].map({'Yes': 1, 'No': 0})
        test_df[col] = test_df[col].map({'Yes': 1, 'No': 0})

    
    # CreditRating
    credit_mapping = {
    '1-Highest': 1, '2-High': 2, '3-Good': 3, '4-Medium': 4, 
    '5-Low': 5, '6-VeryLow': 6, '7-Lowest': 7
    }
    train_df['CreditRating'] = train_df['CreditRating'].map(credit_mapping)
    test_df['CreditRating'] = test_df['CreditRating'].map(credit_mapping)

    # one-hot
    train_df = pd.get_dummies(train_df, columns=['PrizmCode', 'Occupation'], prefix=['Prizm', 'Occ'], dtype=int)
    test_df = pd.get_dummies(test_df, columns=['PrizmCode', 'Occupation'], prefix=['Prizm', 'Occ'], dtype=int)

    return train_df, test_df
    

In [46]:
# Service area 제거
# 비슷한 값을 가지고 있는 PrizmCode가 존재
def preprocess_service_area(train_df, test_df):
    train_df = train_df.drop(columns=['ServiceArea'])
    test_df = test_df.drop(columns=['ServiceArea'])

    return train_df, test_df

In [47]:
# Homeownership 제거
# 필요없음
def preprocess_home_ownership(train_df, test_df):
    train_df = train_df.drop(columns=['Homeownership'])
    test_df = test_df.drop(columns=['Homeownership'])

    return train_df, test_df

In [48]:
def preprocess(train_df, test_df):
    # Churn 전처리
    train_df = preprocess_churn(train_df)
    train_df, test_df = preprocess_encoding(train_df, test_df)
    train_df, test_df = preprocess_service_area(train_df, test_df)
    train_df, test_df = preprocess_home_ownership(train_df, test_df)

    return train_df, test_df


In [49]:
train_df, test_df = preprocess(train_df, test_df)
train_df.head()

,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,Prizm_Suburban,Prizm_Town,Occ_Clerical,Occ_Crafts,Occ_Homemaker,Occ_Other,Occ_Professional,Occ_Retired,Occ_Self,Occ_Student
0,3000002,1,24.00,219.0,22.0,0.25,0.0,0.0,-157.0,-19.0,...,1,0,0,0,0,0,1,0,0,0
1,3000010,1,16.99,10.0,17.0,0.00,0.0,0.0,-4.0,0.0,...,1,0,0,0,0,0,1,0,0,0
2,3000014,0,38.00,8.0,38.0,0.00,0.0,0.0,-2.0,0.0,...,0,1,0,1,0,0,0,0,0,0
3,3000022,0,82.28,1312.0,75.0,1.24,0.0,0.0,157.0,8.1,...,0,0,0,0,0,1,0,0,0,0
4,3000026,1,17.14,0.0,17.0,0.00,0.0,0.0,0.0,-0.2,...,0,0,0,0,0,0,1,0,0,0


## 결측치

In [50]:
train_df.isnull().sum().sort_values(ascending=False)

AgeHH2                909
AgeHH1                909
PercChangeMinutes     367
PercChangeRevenues    367
RoamingCalls          156
                     ... 
Occ_Other               0
Occ_Professional        0
Occ_Retired             0
Occ_Self                0
Occ_Student             0
Length: 66, dtype: int64

In [51]:
test_df.isnull().sum().sort_values(ascending=False)

Churn                 20000
AgeHH1                  335
AgeHH2                  335
PercChangeMinutes       135
PercChangeRevenues      135
                      ...  
Occ_Other                 0
Occ_Professional          0
Occ_Retired               0
Occ_Self                  0
Occ_Student               0
Length: 66, dtype: int64

In [52]:
# train_df의 결측치를 하나라도 가지고 있는 행의 수
missing_rows = train_df.isnull().any(axis=1).sum()
total_rows = len(train_df)

ratio = (missing_rows / total_rows) * 100
print(f'결측치 비율: {ratio:.2f}%')

결측치 비율: 2.49%


In [53]:
# test_df의 결측치를 하나라도 가지고 있는 행의 수
missing_rows = test_df.drop(columns=['Churn']).isnull().any(axis=1).sum()
total_rows = len(test_df)

ratio = (missing_rows / total_rows) * 100
print(f'결측치 비율: {ratio:.2f}%')

결측치 비율: 2.33%


##### 함수 정의

In [54]:
# null이 존재하는 행의 수가 0보다 크고, null이 존재하는 행들의 인덱스가 모두 동일한지 확인
# null이 존재하지 않으면
def check_same_null(df, columns):
    null_mask = df[columns].isnull()
    any_null = null_mask.any(axis=1)
    
    if not any_null.any():
        print("Null 값 없음")
        return

    # 모든 컬럼이 동시에 Null인 행의 수와 하나라도 Null인 행의 수가 같은지 비교
    if any_null.sum() == null_mask.all(axis=1).sum():
        print("Null이 존재하는 행이 모두 일치함")
    else:
        print("컬럼마다 Null 발생 위치가 다름")

In [55]:
def remove_null_rows(df, columns):
    ret_df = df.dropna(subset=columns)
    return ret_df

### Handsets, HandsetModels, CurrentEquipmentDays 처리 (결측치 1개)

In [56]:
# Handsets, HandsetModels, CurrentEquipmentDays는 전부 같은 행에서 결측치를 가짐
# 다른 값으로 대체하기 어려움, 1개 뿐이므로 제거
col1 = ['Handsets', 'HandsetModels', 'CurrentEquipmentDays']
check_same_null(train_df, col1)
check_same_null(test_df, col1)

Null이 존재하는 행이 모두 일치함
Null 값 없음


In [57]:
train_df = remove_null_rows(train_df, col1)
check_same_null(train_df, col1)

Null 값 없음


### PercChangeRevenues, PercChangeMinutes 처리 (결측치 367개)

In [58]:
# PercChangeRevenues, PercChangeMinutes 전부 같은 행에서 결측치를 가짐
# 각각 수익변화율, 사용시간변화율
col2 = ['PercChangeRevenues', 'PercChangeMinutes']
check_same_null(train_df, col2)
check_same_null(test_df, col2)


# 혹시 신규 고객이라 이전 사용량이 없다면??? -> 서비스 이용 기간과 새 단말인지 확인
# -> 평균 이용기간 26.6개월로, 신규고객으로 보기 어려움
display(train_df[train_df['PercChangeRevenues'].isnull()]['MonthsInService'].describe())
display(train_df[train_df['PercChangeRevenues'].isnull()]['NewCellphoneUser'].value_counts())

Null이 존재하는 행이 모두 일치함
Null이 존재하는 행이 모두 일치함


count    367.000000
mean      26.634877
std       13.171533
min        7.000000
25%       16.000000
50%       26.000000
75%       31.500000
max       60.000000
Name: MonthsInService, dtype: float64

NewCellphoneUser
0    331
1     36
Name: count, dtype: int64

In [59]:
# 대체값을 찾기 어려움, 제거
train_df = remove_null_rows(train_df, col2)
test_df = remove_null_rows(test_df, col2)

check_same_null(train_df, col2)
check_same_null(test_df, col2)

Null 값 없음
Null 값 없음


### DirectorAssistedCalls, TotalRecurringCharge. RoamingCalls, OverageMinutes, MonthlyRevenue, MonthlyMinutes 처리 결측치 (156개)

In [60]:
# PercChangeRevenues를 제거하면서 함께 제거됨
# 즉, 이 열에서 결측치를 가지고 있던 행들은 너무 많은 결측치(col2 + col3)를 가지고 있어서 제거될만 했음
col3 = ['DirectorAssistedCalls', 'TotalRecurringCharge', 'RoamingCalls', 'OverageMinutes', 'MonthlyRevenue', 'MonthlyMinutes']
check_same_null(train_df, col3)
check_same_null(test_df, col3)

print(train_df['DirectorAssistedCalls'].isnull().sum())

Null 값 없음
Null 값 없음
0


### AgeHH1, AgeHH2 처리 (결측치 909개)

In [61]:
# AgeHH1, AgeHH2는 전부 같은 행에서 결측치를 가짐
col4 = ['AgeHH1', 'AgeHH2']
check_same_null(train_df, col4)
check_same_null(test_df, col4)

# 1인 가구일 가능성을 고려
train_df[(train_df['AgeHH1'] == 0) & (train_df['AgeHH2'] == 0)]['MaritalStatus'].value_counts()

Null이 존재하는 행이 모두 일치함
Null이 존재하는 행이 모두 일치함


MaritalStatus
Unknown    12261
No          1429
Yes          119
Name: count, dtype: int64

In [62]:
# Unknown이 너무 많아 판단 어려움. No 비율이 더 높긴 하지만, Unknown 자체가 많아서 제거하기로 결정
train_df = remove_null_rows(train_df, col4)
test_df = remove_null_rows(test_df, col4)

In [63]:
display(train_df.isnull().sum().sort_values(ascending=False))
display(test_df.isnull().sum().sort_values(ascending=False))

CustomerID              0
Churn                   0
MonthlyRevenue          0
MonthlyMinutes          0
TotalRecurringCharge    0
                       ..
Occ_Other               0
Occ_Professional        0
Occ_Retired             0
Occ_Self                0
Occ_Student             0
Length: 66, dtype: int64

Churn                   19533
CustomerID                  0
MonthlyRevenue              0
MonthlyMinutes              0
TotalRecurringCharge        0
                        ...  
Occ_Other                   0
Occ_Professional            0
Occ_Retired                 0
Occ_Self                    0
Occ_Student                 0
Length: 66, dtype: int64

### HandsetPrice, MaritalStatus

In [64]:
# HandsetPrice의 Unknown 비율 확인
print(len(train_df[train_df['HandsetPrice'] == 'Unknown']) / len(train_df['HandsetPrice']) * 100)
print(len(test_df[test_df['HandsetPrice'] == 'Unknown']) / len(test_df['HandsetPrice']) * 100)

print("\n")

# MaritalStatus의 Unknown 비율 확인
print(len(train_df[train_df['MaritalStatus'] == 'Unknown']) / len(train_df['MaritalStatus']) * 100)
print(len(test_df[test_df['MaritalStatus'] == 'Unknown']) / len(test_df['MaritalStatus']) * 100)

56.80649308903889
56.32007372139456


37.47388299582128
37.1320329698459


In [65]:
# Unknown값의 비율이 너무 높다
# Unknown -> 조사에 응답하지 않는 고객을 모델이 학습에 이용할 가능성을 열어두고 유지하는 방향으로 결정

# HandsetPrice: 수치형이기 때문에 Median으로 채움(중앙값이 평균보다 이상치에 강하다고 함)
def preprocess_handset_price(train_df, test_df):
    train_df['HandsetPrice'] = pd.to_numeric(train_df['HandsetPrice'].replace('Unknown', np.nan))
    test_df['HandsetPrice'] = pd.to_numeric(test_df['HandsetPrice'].replace('Unknown', np.nan))

    median_price = train_df['HandsetPrice'].median()
    train_df['HandsetPrice'] = train_df['HandsetPrice'].fillna(median_price)
    test_df['HandsetPrice'] = test_df['HandsetPrice'].fillna(median_price)

    return train_df, test_df

# MaritalStatus: 범주형이기 때문에 원핫 인코딩
def preprocess_marital_status(train_df, test_df):
    train_df['MaritalStatus'] = train_df['MaritalStatus'].replace('Unknown', np.nan)
    test_df['MaritalStatus'] = test_df['MaritalStatus'].replace('Unknown', np.nan)

    # 원핫 인코딩
    train_df = pd.get_dummies(train_df, columns=['MaritalStatus'], prefix='Marital', dtype=int)
    test_df = pd.get_dummies(test_df, columns=['MaritalStatus'], prefix='Marital', dtype=int)

    return train_df, test_df

In [66]:
train_df, test_df = preprocess_handset_price(train_df, test_df)
train_df, test_df = preprocess_marital_status(train_df, test_df)

In [67]:
display(train_df.head())
display(test_df.head())

,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,Occ_Clerical,Occ_Crafts,Occ_Homemaker,Occ_Other,Occ_Professional,Occ_Retired,Occ_Self,Occ_Student,Marital_No,Marital_Yes
0,3000002,1,24.00,219.0,22.0,0.25,0.0,0.0,-157.0,-19.0,...,0,0,0,0,1,0,0,0,1,0
1,3000010,1,16.99,10.0,17.0,0.00,0.0,0.0,-4.0,0.0,...,0,0,0,0,1,0,0,0,0,1
2,3000014,0,38.00,8.0,38.0,0.00,0.0,0.0,-2.0,0.0,...,0,1,0,0,0,0,0,0,0,1
3,3000022,0,82.28,1312.0,75.0,1.24,0.0,0.0,157.0,8.1,...,0,0,0,1,0,0,0,0,1,0
4,3000026,1,17.14,0.0,17.0,0.00,0.0,0.0,0.0,-0.2,...,0,0,0,0,1,0,0,0,0,1


,CustomerID,Churn,MonthlyRevenue,MonthlyMinutes,TotalRecurringCharge,DirectorAssistedCalls,OverageMinutes,RoamingCalls,PercChangeMinutes,PercChangeRevenues,...,Occ_Clerical,Occ_Crafts,Occ_Homemaker,Occ_Other,Occ_Professional,Occ_Retired,Occ_Self,Occ_Student,Marital_No,Marital_Yes
0,3000006,NaN,57.49,483.0,37.0,0.25,23.0,0.0,532.0,51.0,...,0,0,0,1,0,0,0,0,1,0
1,3000018,NaN,55.23,570.0,72.0,0.00,0.0,0.0,38.0,0.0,...,0,0,0,0,1,0,0,0,1,0
2,3000034,NaN,97.34,1039.0,50.0,4.95,420.0,0.0,198.0,23.3,...,0,1,0,0,0,0,0,0,0,1
3,3000070,NaN,35.59,153.0,30.0,0.00,16.0,0.0,30.0,7.3,...,0,0,0,1,0,0,0,0,1,0
4,3000074,NaN,55.27,1213.0,50.0,0.74,0.0,1.3,169.0,1.0,...,0,0,0,1,0,0,0,0,1,0


### 이후 모델 학습을 용이하게 하기 위해 'Churn'열을 제거하고 따로 저장 

In [68]:
# 이후 모델 학습을 용이하게 하기 위해 'Churn'열을 제거하고 따로 저장 (모델 학습 시 X, y로 나누기 위해)
def preprocess_remove_churn(train_df, test_df):
    train_churn = train_df['Churn']
    train_df = train_df.drop(columns=['Churn'])
    test_df = test_df.drop(columns=['Churn'])

    return train_df, test_df, train_churn